File: src/python/core/application/services/skill_document_builder.py
from typing import List, Dict, Any
from langchain_core.documents import Document

from ...domain.entities.candidate_record import CandidateRecord
from ...domain.enums.skill_level import SkillLevel
from ...domain.configuration.vector_metadata_config import VectorMetadataConfig
from .skill_normalizer import SkillNormalizer


class SkillDocumentBuilder:
    def __init__(self, config: VectorMetadataConfig):
        self._config = config
    
    def build_skill_documents(
        self,
        candidate: CandidateRecord,
        candidate_id: str,
        seniority_level: str,
        years_experience: int
    ) -> List[Document]:
        documents = []
        
        if not candidate.SkillMatrix:
            return documents
        
        fullname = None
        if candidate.GeneralInfo and candidate.GeneralInfo.Fullname and candidate.GeneralInfo.Fullname.strip():
            extracted_fullname = candidate.GeneralInfo.Fullname.strip()
            if extracted_fullname != candidate_id:
                fullname = extracted_fullname
            else:
                print(f"[WARNING] Fullname equals candidate_id for {candidate_id}, NOT storing in skill metadata")
        
        for skill in candidate.SkillMatrix:
            if not skill.SkillName or not skill.SkillLevel:
                continue
            
            skill_level_enum = SkillLevel.from_string(skill.SkillLevel)
            
            if not SkillLevel.is_strong(skill_level_enum):
                continue
            
            normalized_skill_name = SkillNormalizer.normalize(skill.SkillName)
            
            metadata = self._build_skill_metadata(
                candidate_id=candidate_id,
                fullname=fullname,
                skill_name=normalized_skill_name,
                skill_level=skill.SkillLevel,
                seniority_level=seniority_level,
                years_experience=years_experience
            )
            
            page_content = self._build_skill_content(
                skill_name=skill.SkillName,
                skill_level=skill.SkillLevel,
                evidence=skill.Evidence
            )
            
            documents.append(Document(
                page_content=page_content,
                metadata=metadata
            ))
        
        return documents
    
    def _build_skill_metadata(
        self,
        candidate_id: str,
        fullname: str,
        skill_name: str,
        skill_level: str,
        seniority_level: str,
        years_experience: int
    ) -> Dict[str, Any]:
        metadata = {
            self._config.FIELD_TYPE: self._config.TYPE_SKILL,
            self._config.FIELD_CANDIDATE_ID: candidate_id,
            self._config.FIELD_SKILL_NAME: skill_name,
            self._config.FIELD_SKILL_LEVEL: skill_level,
            self._config.FIELD_SENIORITY_LEVEL: seniority_level,
            self._config.FIELD_YEARS_EXPERIENCE: years_experience
        }
        
        if fullname:
            metadata[self._config.FIELD_FULLNAME] = fullname
        
        return metadata
    
    def _build_skill_content(
        self,
        skill_name: str,
        skill_level: str,
        evidence: str = None
    ) -> str:
        content = f"{skill_name} ({skill_level})"
        
        if evidence:
            content += f": {evidence}"
        
        return content

File: src/dotnet/Core/Application/Services/SkillDocumentBuilder.cs
using Rag.Candidates.Core.Domain.Configuration;
using Rag.Candidates.Core.Domain.Entities;
using Rag.Candidates.Core.Domain.Enums;
using Microsoft.SemanticKernel.Memory;

namespace Rag.Candidates.Core.Application.Services;

public sealed class SkillDocumentBuilder
{
    private readonly VectorMetadataConfig _config;

    public SkillDocumentBuilder(VectorMetadataConfig config)
    {
        _config = config;
    }

    public List<(string Content, Dictionary<string, object> Metadata)> BuildSkillDocuments(
        CandidateRecord candidate,
        string candidateId,
        string seniorityLevel,
        int yearsExperience)
    {
        var documents = new List<(string, Dictionary<string, object>)>();

        if (candidate.SkillMatrix == null)
            return documents;

        // Only extract fullname if it exists and is not empty
        // NEVER use candidateId as fullname
        string? fullname = null;
        if (!string.IsNullOrWhiteSpace(candidate.GeneralInfo?.Fullname))
        {
            var extractedFullname = candidate.GeneralInfo.Fullname.Trim();
            if (extractedFullname != candidateId)
            {
                fullname = extractedFullname;
            }
            else
            {
                Console.WriteLine($"[WARNING] Fullname equals candidateId for {candidateId}, NOT storing in skill metadata");
            }
        }

        foreach (var skill in candidate.SkillMatrix)
        {
            if (string.IsNullOrWhiteSpace(skill.Name) || string.IsNullOrWhiteSpace(skill.Level))
                continue;

            if (!SkillLevelExtensions.IsStrong(skill.Level))
                continue;

            var normalizedSkillName = SkillNormalizer.Normalize(skill.Name);

            var metadata = BuildSkillMetadata(
                candidateId: candidateId,
                fullname: fullname,
                skillName: normalizedSkillName,
                skillLevel: skill.Level,
                seniorityLevel: seniorityLevel,
                yearsExperience: yearsExperience
            );

            var content = BuildSkillContent(
                skillName: skill.Name,
                skillLevel: skill.Level,
                evidence: skill.Evidence
            );

            documents.Add((content, metadata));
        }

        return documents;
    }

    private Dictionary<string, object> BuildSkillMetadata(
        string candidateId,
        string? fullname,
        string skillName,
        string skillLevel,
        string seniorityLevel,
        int yearsExperience)
    {
        var metadata = new Dictionary<string, object>
        {
            [_config.FieldType] = _config.TypeSkill,
            [_config.FieldCandidateId] = candidateId,
            [_config.FieldSkillName] = skillName,
            [_config.FieldSkillLevel] = skillLevel,
            [_config.FieldSeniorityLevel] = seniorityLevel,
            [_config.FieldYearsExperience] = yearsExperience
        };
        
        // Only add fullname if it exists (not null)
        if (fullname != null)
        {
            metadata[_config.FieldFullname] = fullname;
        }
        
        return metadata;
    }

    private static string BuildSkillContent(
        string skillName,
        string skillLevel,
        string? evidence = null)
    {
        var content = $"{skillName} ({skillLevel})";

        if (!string.IsNullOrWhiteSpace(evidence))
        {
            content += $": {evidence}";
        }

        return content;
    }
}
